### ISSUE TITLE
RoC Liquid Pipes with Liquid XP spam NPEs client-side

### ISSUE THREAD

**Issue by thePalindrome** [ReikaKalseki/Reika_Mods_Issues#2197]
https://paste.ee/p/t3l7y

Any tick a liquid pipe with Liquid XP, it'll spam chat and the log with client-side NPEs (they don't cause crashing, they just cause massive spam to the chat and log, rendering both nigh unusable).

This effect duplicates for each liquid pipe with xp in it, and since those seem to be not liking EIO conduits (see #2196 ), I'm kindof stuck with the console spam/not being able to automatically withdraw the liquid xp (needed for automatic ChrC chroma production)

**Comment by ReikaKalseki** [comment]
This is very strange.

    [23:19:39] [Server thread/ERROR] [FML]: DRAGONAPI ERROR: Tile Entity Liquid Pipe @ DIM0: -2340, 69, 2788 [SERVER] is throwing class java.lang.NullPointerException on update: null
    [23:19:39] [Server thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: java.lang.NullPointerException
    [23:19:39] [Server thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at Reika.RotaryCraft.Base.TileEntity.TileEntityPiping.intakeFluid(TileEntityPiping.java:402)
    [23:19:39] [Server thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at Reika.RotaryCraft.Base.TileEntity.TileEntityPiping.updateEntity(TileEntityPiping.java:140)
    [23:19:39] [Server thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at Reika.RotaryCraft.TileEntities.Piping.TileEntityPipe.updateEntity(TileEntityPipe.java:46)
    [23:19:39] [Server thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at Reika.DragonAPI.Base.TileEntityBase.func_145845_h(TileEntityBase.java:510)
    [23:19:39] [Server thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.world.World.func_72939_s(World.java:1939)
    [23:19:39] [Server thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.world.WorldServer.func_72939_s(WorldServer.java:489)
    [23:19:39] [Server thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.server.MinecraftServer.func_71190_q(MinecraftServer.java:636)
    [23:19:39] [Server thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.server.MinecraftServer.func_71217_p(MinecraftServer.java:547)
    [23:19:39] [Server thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.server.integrated.IntegratedServer.func_71217_p(IntegratedServer.java:111)
    [23:19:39] [Server thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.server.MinecraftServer.run(MinecraftServer.java:427)
    [23:19:39] [Server thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.server.MinecraftServer$2.run(MinecraftServer.java:685)

**Comment by thePalindrome** [comment]
Attaching to this issue since I *think* it may be related. I was testing things in my creative world, and my Reactorcraft turbines were giving this so long as they were running (as in, if I removed the steam source it stops)

```
[01:26:53] [Client thread/ERROR] [FML]: DRAGONAPI ERROR: Tile Entity Turbine @ DIM0: -325, 9, 305 [CLIENT] is throwing class java.lang.NullPointerException on update: null
[01:26:53] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: java.lang.NullPointerException
[01:26:53] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at Reika.ReactorCraft.TileEntities.PowerGen.TileEntityTurbineCore.intakeSteam(TileEntityTurbineCore.java:450)
[01:26:53] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at Reika.ReactorCraft.TileEntities.PowerGen.TileEntityTurbineCore.readSurroundings(TileEntityTurbineCore.java:500)
[01:26:53] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at Reika.ReactorCraft.TileEntities.PowerGen.TileEntityTurbineCore.updateEntity(TileEntityTurbineCore.java:164)
[01:26:53] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at Reika.DragonAPI.Base.TileEntityBase.func_145845_h(TileEntityBase.java:510)
[01:26:53] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.world.World.func_72939_s(World.java:1939)
[01:26:53] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.client.Minecraft.func_71407_l(Minecraft.java:2006)
[01:26:53] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.client.Minecraft.func_71411_J(Minecraft.java:973)
[01:26:53] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.client.Minecraft.func_99999_d(Minecraft.java:898)
[01:26:53] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.client.main.Main.main(SourceFile:148)
[01:26:53] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[01:26:53] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[01:26:53] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[01:26:53] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at java.lang.reflect.Method.invoke(Method.java:498)
[01:26:53] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.launchwrapper.Launch.launch(Launch.java:135)
[01:26:53] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at net.minecraft.launchwrapper.Launch.main(Launch.java:28)
[01:26:53] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[01:26:53] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[01:26:53] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[01:26:53] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at java.lang.reflect.Method.invoke(Method.java:498)
[01:26:53] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at org.multimc.onesix.OneSixLauncher.launchWithMainClass(OneSixLauncher.java:196)
[01:26:53] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at org.multimc.onesix.OneSixLauncher.launch(OneSixLauncher.java:231)
[01:26:53] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at org.multimc.EntryPoint.listen(EntryPoint.java:143)
[01:26:53] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:749]: 	at org.multimc.EntryPoint.main(EntryPoint.java:34)
[01:26:53] [Client thread/INFO] [FML]: DRAGONAPI: 
[01:26:53] [Client thread/INFO]: [CHAT] Tile Entity Turbine @ DIM0: -325, 9, 305 [CLIENT] is throwing class java.lang.NullPointerException on update: null
```


Given the stacktrace on that, I don't think it's related, but figured it'd be worth attaching here (I can make a seperate issue if it's not helpful to this)

**Comment by ReikaKalseki** [comment]
I highly doubt it to be related.

**Comment by ReikaKalseki** [comment]
This looks very much like it is not my doing.

The line throwing the error is
`int drained = fl.drain(dir.UNKNOWN, todrain, true).amount;`
which can only NPE on the result of `fl.drain()` (fl is queried earlier and thus cannot be null).

However, there should be no way that `fl.drain()` should return null; That line is only executed after `FluidStack fs = fl.drain(dir.UNKNOWN, Integer.MAX_VALUE, false);`, and only if fs is not null.

In other words, this error depends on the IFluidHandler tile first providing a non-null drain result when simulating a drain, but then returning null during the actual drain. This is a violation of how those methods are supposed to work.

**Comment by thePalindrome** [comment]
That does seem really really weird, especially since it's only *your* pipes doing that on *your* machine. I don't *think* I have any other mods that would be interfering with it... Would there be any possibility that you could make some debug tweaks or something?

Oh! Looking back at this issue, I forgot to mention a detail (I only mentioned it on the other one):
The Thermal Dynamics Fluiducts don't seem to do that!

**Comment by ReikaKalseki** [comment]
They have different fluid logic, so that is no surprise.

**Comment by habnabit** [comment]
@thePalindrome the turbine issue is #2147. (Not related to optifine, despite the title.)
